File: app/src/main/java/com/projectfox/foxoff/brain/FoxBrainEvent.kt
package com.projectfox.foxoff.brain

import java.time.Instant

/**
 * Represent all possible inputs for the Fox Brain engine.
 */
sealed class FoxBrainEvent {
    // `internal set` : permet à NightSimulator (app/src/test, voir
    // brain/simulation/) de rejouer une nuit entière avec des horodatages
    // fictifs (23h-7h) sans attendre 8h en temps réel, sans toucher à la
    // valeur par défaut ni à aucun site d'appel existant.
    var timestamp: Instant = Instant.now()
        internal set

    // Heart Rate Events
    data class HeartRateReceived(val bpm: Float, val source: String) : FoxBrainEvent()
    data class PassiveHeartRateReceived(val bpm: Float) : FoxBrainEvent()
    data class ExerciseHeartRateReceived(val bpm: Float) : FoxBrainEvent()

    // Movement Events
    data class MovementDetected(val magnitude: Float) : FoxBrainEvent()
    object StillnessDetected : FoxBrainEvent()

    // TV Events
    object TVTurnedOn : FoxBrainEvent()
    object TVTurnedOff : FoxBrainEvent()
    data class TvAppChanged(val appName: String?) : FoxBrainEvent()
    data class TvCommandSent(val command: String, val time: Instant) : FoxBrainEvent()

    // Device / Watch Events
    data class WatchConnected(val name: String) : FoxBrainEvent()
    object WatchDisconnected : FoxBrainEvent()
    data class BatteryChanged(val level: Int, val isCharging: Boolean) : FoxBrainEvent()

    // Phone / User Interface Events
    object ScreenLocked : FoxBrainEvent()
    object ScreenUnlocked : FoxBrainEvent()

    // Sleep Events
    object SleepDetected : FoxBrainEvent()
    object SleepCancelled : FoxBrainEvent()
    object WalkingDetected : FoxBrainEvent()

    // System Events
    data class TimeChanged(val localTime: Instant) : FoxBrainEvent()
    object MonitoringStarted : FoxBrainEvent()
    object MonitoringStopped : FoxBrainEvent()
}

File: ios/FoxCoreKit/Sources/FoxCoreKit/FoxBrainEvent.swift
import Foundation

/// Portage de `com.projectfox.foxoff.brain.FoxBrainEvent` (Kotlin, `sealed
/// class`). Kotlin donne un timestamp par défaut à chaque sous-type via un
/// champ commun de la classe scellée ; Swift n'a pas d'équivalent direct
/// pour un `enum` à cas associés, d'où cette enveloppe `FoxBrainEvent` qui
/// porte le timestamp à côté du `payload` (le vrai contenu de l'événement).
public struct FoxBrainEvent {
    public let timestamp: Date
    public let payload: Payload

    public init(_ payload: Payload, timestamp: Date = Date()) {
        self.payload = payload
        self.timestamp = timestamp
    }

    public enum Payload {
        // Événements de fréquence cardiaque
        case heartRateReceived(bpm: Float, source: String)
        case passiveHeartRateReceived(bpm: Float)
        case exerciseHeartRateReceived(bpm: Float)

        // Événements de mouvement
        case movementDetected(magnitude: Float)
        case stillnessDetected

        // Événements TV
        case tvTurnedOn
        case tvTurnedOff
        case tvAppChanged(appName: String?)
        case tvCommandSent(command: String, time: Date)

        // Événements montre
        case watchConnected(name: String)
        case watchDisconnected
        case batteryChanged(level: Int, isCharging: Bool)

        // Événements téléphone / interface utilisateur
        case screenLocked
        case screenUnlocked

        // Événements de sommeil
        case sleepDetected
        case sleepCancelled
        case walkingDetected

        // Événements système
        case timeChanged(localTime: Date)
        case monitoringStarted
        case monitoringStopped
    }
}
